### ISSUE TITLE
make test and make bench coredump on aarch64

### ISSUE THREAD

**Issue by baiwfg2** [rafaelkallis/adaptive-radix-tree#17]
Hi. I build it with gcc 7.3 on aarch64 linux. And `make test` and `make bench` don't work out.
```
root@adaptive-radix-tree (master)$ make test
[doctest] doctest version is "2.3.8"
[doctest] run with "--help" for options
terminate called after throwing an instance of 'std::out_of_range'
  what():  provided partial key does not have a successor
===============================================================================
/data/adaptive-radix-tree/test/art.cpp:25:
TEST SUITE: art
TEST CASE:  set

DEEPEST SUBCASE STACK REACHED (DIFFERENT FROM THE CURRENT ONE):
  insert value s.t. existing value is a prefix

/data/adaptive-radix-tree/test/art.cpp:25: FATAL ERROR: test case CRASHED: SIGABRT - Abort (abnormal termination) signal                                                                

===============================================================================
[doctest] test cases:      1 |      0 passed |      1 failed |     23 skipped
[doctest] assertions:      5 |      5 passed |      0 failed |
[doctest] Status: FAILURE!
/bin/sh: line 1:  3308 Aborted                 (core dumped) ./build/test
make: *** [Makefile:14: test] Error 134

root@adaptive-radix-tree (master)$ make bench
terminate called after throwing an instance of 'std::out_of_range'
  what():  provided partial key does not have a successor
/bin/sh: line 1:  4235 Aborted                 (core dumped) ./build/bench
make: *** [Makefile:17: bench] Error 134

```

**Comment by rafaelkallis** [comment]
Hi @baiwfg2, could you test the above in the `preview` branch and see if the error happens there as well?

**Comment by baiwfg2** [comment]
It seems to be same problem.

```
root@adaptive-radix-tree (preview)$ make test
[doctest] doctest version is "2.3.8"
[doctest] run with "--help" for options
terminate called after throwing an instance of 'std::out_of_range'
  what():  provided partial key does not have a successor
===============================================================================
/data/adaptive-radix-tree/test/art.cpp:26:
TEST SUITE: art
TEST CASE:  set

DEEPEST SUBCASE STACK REACHED (DIFFERENT FROM THE CURRENT ONE):
  insert value s.t. existing value is a prefix

/data/adaptive-radix-tree/test/art.cpp:26: FATAL ERROR: test case CRASHED: SIGABRT - Abort (abnormal termination) signal

===============================================================================
[doctest] test cases:      1 |      0 passed |      1 failed |     23 skipped
[doctest] assertions:      5 |      5 passed |      0 failed |
[doctest] Status: FAILURE!
/bin/sh: line 1: 29591 Aborted                 (core dumped) ./build/test
make: *** [Makefile:14: test] Error 134

```

**Comment by baiwfg2** [comment]
Here is the debug-version backtrace:
```
(gdb) r
Starting program: /data/adaptive-radix-tree/build/test 
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib64/libthread_db.so.1".
[doctest] doctest version is "2.3.8"
[doctest] run with "--help" for options
terminate called after throwing an instance of 'std::out_of_range'
  what():  provided partial key does not have a successor

Program received signal SIGABRT, Aborted.
0x0000ffffbf2b0140 in raise () from /lib64/libc.so.6
(gdb) bt
#0  0x0000ffffbf2b0140 in raise () from /lib64/libc.so.6
#1  0x0000ffffbf2b14ec in abort () from /lib64/libc.so.6
#2  0x0000ffffbf5981d0 in __gnu_cxx::__verbose_terminate_handler() () from /lib64/libstdc++.so.6
#3  0x0000ffffbf595ae4 in ?? () from /lib64/libstdc++.so.6
#4  0x0000ffffbf594928 in ?? () from /lib64/libstdc++.so.6
#5  0x0000ffffbf595368 in __gxx_personality_v0 () from /lib64/libstdc++.so.6
#6  0x0000ffffbf410e18 in ?? () from /lib64/libgcc_s.so.1
#7  0x0000ffffbf4111bc in _Unwind_RaiseException () from /lib64/libgcc_s.so.1
#8  0x0000ffffbf595de0 in __cxa_throw () from /lib64/libstdc++.so.6
#9  0x00000000004497a8 in art::node_4<int*>::next_partial_key (this=0x4f6ac0, partial_key=128 '\200') at /data/adaptive-radix-tree/include/art/node_4.hpp:115
#10 0x0000000000447ad0 in art::child_it<int*>::child_it (this=0xffffffff35d0, n=0x4f6ac0, relative_index=0) at /data/adaptive-radix-tree/include/art/child_it.hpp:74
#11 0x0000000000447998 in art::child_it<int*>::child_it (this=0xffffffff35d0, n=0x4f6ac0) at /data/adaptive-radix-tree/include/art/child_it.hpp:54
#12 0x0000000000446a38 in art::inner_node<int*>::begin (this=0x4f6ac0) at /data/adaptive-radix-tree/include/art/inner_node.hpp:115
#13 0x00000000004450f8 in art::art<int*>::~art (this=0xffffffffdd40, __in_chrg=<optimized out>) at /data/adaptive-radix-tree/include/art/art.hpp:89
#14 0x000000000043cdb4 in _DOCTEST_ANON_SUITE_2::_DOCTEST_ANON_FUNC_3 () at /data/adaptive-radix-tree/test/art.cpp:28
#15 0x000000000045a5e0 in doctest::Context::run (this=0xffffffffe5e8) at /data/adaptive-radix-tree/third_party/doctest/doctest/doctest.h:5907
#16 0x000000000045adec in main (argc=1, argv=0xffffffffe728) at /data/adaptive-radix-tree/third_party/doctest/doctest/doctest.h:5991
```

**Comment by Meandres** [comment]
For anyone encountering the same problem, allow me to save you the day I spent debugging this.

TLDR: `char` is considered unsigned on aarch64 but signed on x86_64. Compile with `-fsigned-char`.

So the bug actually happens when scanning the tree/when calling `begin()` on the tree. 
Internally it creates a `tree_it` that uses, at some point, a `child_it`. This `child_it` will iterate over a node using the partial_key, for example [here](https://github.com/rafaelkallis/adaptive-radix-tree/blob/master/include/art/child_it.hpp#L74). However, the `partial_key` is of type `char` (see [here](https://github.com/rafaelkallis/adaptive-radix-tree/blob/master/include/art/node_4.hpp#L108) for `node_4`). 
Turns out that aarch64 consider `char` as unsigned whereas x64 considers it as signed. Here is a simple program to verify it:
```
#include <cstdio>
int main(){
    printf("char: %i\n", (char)-1);
    printf("signed char: %i\n", (signed char)-1);
    printf("unsigned char: %i\n", (unsigned char)-1);
}
```
This program outputs the following on aarch64:
```
char: 255
signed char: -1
unsigned char: 255
```
and this on x86_64:
```
char: -1
signed: -1
unsigned: 255
```
**regardless of the compiler** (I tried with clang 19.1.7 and GCC 14.3.0)

So because the sweep starts at -128, the aarch64 version skips over all possible (valid) char values and reports that it didn't find anything which causes the coredump following this [exception](https://github.com/rafaelkallis/adaptive-radix-tree/blob/master/include/art/node_4.hpp#L115)

To fix it, either modify the sweep to start at 0 or compile using `-fsigned-char` (on GCC) to force the compiler to use signed char by default.

**Comment by rafaelkallis** [comment]
> To fix it, either modify the sweep to start at 0 or compile using -fsigned-char (on GCC) to force the compiler to use signed char by default.

@Meandres many thanks for your help. As a token of gratitude, let me buy you a beer. If you share your revolut handle I'm happy to send a tip over, just send me an email. 

If you send me your github account email address I also would be happy to add you as a commit co-author, so far I only could find your academic email.

**Comment by Meandres** [comment]
Haha, no problem, I needed it for my project. Thanks for open-sourcing the implementation in the first place.
